File: go/pkg/compositions/work.go
package compositions

import (
	"context"
	"fmt"

	"github.com/transpara-ai/eventgraph/go/pkg/event"
	"github.com/transpara-ai/eventgraph/go/pkg/grammar"
	"github.com/transpara-ai/eventgraph/go/pkg/types"
)

// WorkGrammar provides Layer 1 (Agency) composition operations.
// 12 operations + 3 named functions for task management where AI agents
// and humans operate on the same graph.
type WorkGrammar struct {
	g *grammar.Grammar
}

// NewWorkGrammar creates a WorkGrammar bound to the given base grammar.
func NewWorkGrammar(g *grammar.Grammar) *WorkGrammar {
	return &WorkGrammar{g: g}
}

// --- Operations (12) ---

// Intend declares a desired outcome. (Intent + Emit)
func (w *WorkGrammar) Intend(
	ctx context.Context, source types.ActorID, goal string,
	causes []types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Emit(ctx, source, "intend: "+goal, convID, causes, signer)
}

// Decompose breaks intent into actionable steps. (Choice + Derive)
func (w *WorkGrammar) Decompose(
	ctx context.Context, source types.ActorID, subtask string,
	goal types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Derive(ctx, source, "decompose: "+subtask, goal, convID, signer)
}

// Assign gives work to a specific actor. (Commitment + Delegate)
func (w *WorkGrammar) Assign(
	ctx context.Context, source types.ActorID, assignee types.ActorID,
	scope types.DomainScope, weight types.Weight,
	cause types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Delegate(ctx, source, assignee, scope, weight, cause, convID, signer)
}

// Claim takes on unassigned work. (Intent + Emit)
func (w *WorkGrammar) Claim(
	ctx context.Context, source types.ActorID, work string,
	causes []types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Emit(ctx, source, "claim: "+work, convID, causes, signer)
}

// Prioritize ranks work by importance. (Value + Annotate)
func (w *WorkGrammar) Prioritize(
	ctx context.Context, source types.ActorID, target types.EventID,
	priority string, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Annotate(ctx, source, target, "priority", priority, convID, signer)
}

// Block flags work that cannot proceed. (Risk + Annotate)
func (w *WorkGrammar) Block(
	ctx context.Context, source types.ActorID, target types.EventID,
	blocker string, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Annotate(ctx, source, target, "blocked", blocker, convID, signer)
}

// Unblock removes an impediment to work. (Consequence + Emit)
func (w *WorkGrammar) Unblock(
	ctx context.Context, source types.ActorID, resolution string,
	causes []types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Emit(ctx, source, "unblock: "+resolution, convID, causes, signer)
}

// Progress reports incremental advancement. (Act + Extend)
func (w *WorkGrammar) Progress(
	ctx context.Context, source types.ActorID, update string,
	previous types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Extend(ctx, source, "progress: "+update, previous, convID, signer)
}

// Complete marks work as done with evidence. (Consequence + Emit)
func (w *WorkGrammar) Complete(
	ctx context.Context, source types.ActorID, summary string,
	causes []types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Emit(ctx, source, "complete: "+summary, convID, causes, signer)
}

// Handoff transfers work between actors. (Signal + Consent)
func (w *WorkGrammar) Handoff(
	ctx context.Context, from types.ActorID, to types.ActorID,
	description string, scope types.DomainScope,
	cause types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Consent(ctx, from, to, "handoff: "+description, scope, cause, convID, signer)
}

// Scope defines what an actor may do autonomously. (Capacity + Delegate)
func (w *WorkGrammar) Scope(
	ctx context.Context, source types.ActorID, target types.ActorID,
	scope types.DomainScope, weight types.Weight,
	cause types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Delegate(ctx, source, target, scope, weight, cause, convID, signer)
}

// Review evaluates completed work. (Consequence + Respond)
func (w *WorkGrammar) Review(
	ctx context.Context, source types.ActorID, assessment string,
	target types.EventID, convID types.ConversationID, signer event.Signer,
) (event.Event, error) {
	return w.g.Respond(ctx, source, "review: "+assessment, target, convID, signer)
}

// --- Named Functions (6) ---

// StandupResult holds the events produced by a Standup.
type StandupResult struct {
	Updates   []event.Event
	Priority  event.Event
}

// Standup gathers status from all participants and sets priority:
// Progress (batch) + Prioritize.
func (w *WorkGrammar) Standup(
	ctx context.Context, participants []types.ActorID, updates []string,
	lead types.ActorID, priority string,
	causes []types.EventID, convID types.ConversationID, signer event.Signer,
) (StandupResult, error) {
	if len(participants) != len(updates) {
		return StandupResult{}, fmt.Errorf("standup: participants and updates must have equal length")
	}

	result := StandupResult{}
	var lastID types.EventID
	for i, actor := range participants {
		var prev types.EventID
		if i == 0 {
			if len(causes) > 0 {
				prev = causes[0]
			}
		} else {
			prev = lastID
		}
		progress, err := w.Progress(ctx, actor, updates[i], prev, convID, signer)
		if err != nil {
			return StandupResult{}, fmt.Errorf("standup/progress[%d]: %w", i, err)
		}
		result.Updates = append(result.Updates, progress)
		lastID = progress.ID()
	}

	prio, err := w.Prioritize(ctx, lead, lastID, priority, convID, signer)
	if err != nil {
		return StandupResult{}, fmt.Errorf("standup/prioritize: %w", err)
	}
	result.Priority = prio

	return result, nil
}

// RetrospectiveResult holds the events produced by a Retrospective.
type RetrospectiveResult struct {
	Reviews     []event.Event
	Improvement event.Event
}

// Retrospective reviews work and identifies improvements: Review (batch) + Intend.
func (w *WorkGrammar) Retrospective(
	ctx context.Context, reviewers []types.ActorID, assessments []string,
	lead types.ActorID, improvement string,
	target types.EventID, convID types.ConversationID, signer event.Signer,
) (RetrospectiveResult, error) {
	if len(reviewers) != len(assessments) {
		return RetrospectiveResult{}, fmt.Errorf("retrospective: reviewers and assessments must have equal length")
	}

	result := RetrospectiveResult{}
	reviewIDs := make([]types.EventID, 0, len(reviewers))
	for i, reviewer := range reviewers {
		rev, err := w.Review(ctx, reviewer, assessments[i], target, convID, signer)
		if err != nil {
			return RetrospectiveResult{}, fmt.Errorf("retrospective/review[%d]: %w", i, err)
		}
		result.Reviews = append(result.Reviews, rev)
		reviewIDs = append(reviewIDs, rev.ID())
	}

	improve, err := w.Intend(ctx, lead, improvement, reviewIDs, convID, signer)
	if err != nil {
		return RetrospectiveResult{}, fmt.Errorf("retrospective/intend: %w", err)
	}
	result.Improvement = improve

	return result, nil
}

// TriageResult holds the events produced by a Triage.
type TriageResult struct {
	Priorities  []event.Event
	Assignments []event.Event
	Scopes      []event.Event
}

// Triage prioritises and assigns a batch of items: Prioritize + Assign + Scope (batch).
func (w *WorkGrammar) Triage(
	ctx context.Context, lead types.ActorID,
	items []types.EventID, priorities []string,
	assignees []types.ActorID, scopes []types.DomainScope, weights []types.Weight,
	convID types.ConversationID, signer event.Signer,
) (TriageResult, error) {
	n := len(items)
	if len(priorities) != n || len(assignees) != n || len(scopes) != n || len(weights) != n {
		return TriageResult{}, fmt.Errorf("triage: all slices must have equal length")
	}

	result := TriageResult{}
	for i := range items {
		prio, err := w.Prioritize(ctx, lead, items[i], priorities[i], convID, signer)
		if err != nil {
			return TriageResult{}, fmt.Errorf("triage/prioritize[%d]: %w", i, err)
		}
		result.Priorities = append(result.Priorities, prio)

		assign, err := w.Assign(ctx, lead, assignees[i], scopes[i], weights[i], prio.ID(), convID, signer)
		if err != nil {
			return TriageResult{}, fmt.Errorf("triage/assign[%d]: %w", i, err)
		}
		result.Assignments = append(result.Assignments, assign)

		scope, err := w.Scope(ctx, lead, assignees[i], scopes[i], weights[i], assign.ID(), convID, signer)
		if err != nil {
			return TriageResult{}, fmt.Errorf("triage/scope[%d]: %w", i, err)
		}
		result.Scopes = append(result.Scopes, scope)
	}

	return result, nil
}

// SprintResult holds the events produced by a Sprint.
type SprintResult struct {
	Intent        event.Event
	Subtasks      []event.Event
	Assignments   []event.Event
}

// Sprint plans a work cycle: Intend + Decompose + Assign (batch).
func (w *WorkGrammar) Sprint(
	ctx context.Context, source types.ActorID, goal string,
	subtasks []string, assignees []types.ActorID, scopes []types.DomainScope,
	causes []types.EventID, convID types.ConversationID, signer event.Signer,
) (SprintResult, error) {
	if len(subtasks) != len(assignees) || len(subtasks) != len(scopes) {
		return SprintResult{}, fmt.Errorf("sprint: subtasks, assignees, and scopes must have equal length")
	}

	intent, err := w.Intend(ctx, source, goal, causes, convID, signer)
	if err != nil {
		return SprintResult{}, fmt.Errorf("sprint/intend: %w", err)
	}

	result := SprintResult{Intent: intent}
	for i, st := range subtasks {
		task, err := w.Decompose(ctx, source, st, intent.ID(), convID, signer)
		if err != nil {
			return SprintResult{}, fmt.Errorf("sprint/decompose[%d]: %w", i, err)
		}
		result.Subtasks = append(result.Subtasks, task)

		assign, err := w.Assign(ctx, source, assignees[i], scopes[i], types.MustWeight(0.5), task.ID(), convID, signer)
		if err != nil {
			return SprintResult{}, fmt.Errorf("sprint/assign[%d]: %w", i, err)
		}
		result.Assignments = append(result.Assignments, assign)
	}
	return result, nil
}

// EscalateResult holds the events produced by an Escalate.
type EscalateResult struct {
	BlockEvent   event.Event
	HandoffEvent event.Event
}

// Escalate moves stuck work up: Block + Handoff (to higher authority).
func (w *WorkGrammar) Escalate(
	ctx context.Context, source types.ActorID, blocker string,
	task types.EventID, authority types.ActorID, scope types.DomainScope,
	convID types.ConversationID, signer event.Signer,
) (EscalateResult, error) {
	block, err := w.Block(ctx, source, task, blocker, convID, signer)
	if err != nil {
		return EscalateResult{}, fmt.Errorf("escalate/block: %w", err)
	}

	handoff, err := w.Handoff(ctx, source, authority, blocker, scope, block.ID(), convID, signer)
	if err != nil {
		return EscalateResult{}, fmt.Errorf("escalate/handoff: %w", err)
	}

	return EscalateResult{BlockEvent: block, HandoffEvent: handoff}, nil
}

// DelegateAndVerifyResult holds the events produced by DelegateAndVerify.
type DelegateAndVerifyResult struct {
	AssignEvent event.Event
	ScopeEvent  event.Event
}

// DelegateAndVerify is a full delegation cycle: Assign + Scope.
// The Review step happens later when work is complete.
func (w *WorkGrammar) DelegateAndVerify(
	ctx context.Context, source types.ActorID, assignee types.ActorID,
	scope types.DomainScope, weight types.Weight,
	cause types.EventID, convID types.ConversationID, signer event.Signer,
) (DelegateAndVerifyResult, error) {
	assign, err := w.Assign(ctx, source, assignee, scope, weight, cause, convID, signer)
	if err != nil {
		return DelegateAndVerifyResult{}, fmt.Errorf("delegate-and-verify/assign: %w", err)
	}

	scopeEv, err := w.Scope(ctx, source, assignee, scope, weight, assign.ID(), convID, signer)
	if err != nil {
		return DelegateAndVerifyResult{}, fmt.Errorf("delegate-and-verify/scope: %w", err)
	}

	return DelegateAndVerifyResult{AssignEvent: assign, ScopeEvent: scopeEv}, nil
}

File: rust/src/compositions/work.rs
//! Layer 1 (Agency) composition operations.
//!
//! 12 operations + 6 named functions for task management where AI agents
//! and humans operate on the same graph.

use crate::errors::Result;
use crate::event::{Event, Signer};
use crate::grammar::Grammar;
use crate::store::InMemoryStore;
use crate::types::{ActorId, ConversationId, DomainScope, EventId, Weight};

/// WorkGrammar provides Layer 1 (Agency) composition operations.
pub struct WorkGrammar<'a>(Grammar<'a>);

impl<'a> WorkGrammar<'a> {
    pub fn new(store: &'a mut InMemoryStore) -> Self {
        Self(Grammar::new(store))
    }

    // --- Operations (12) ---

    /// Intend declares intent or desired outcome. (Intent + Emit)
    pub fn intend(
        &mut self,
        source: ActorId,
        goal: &str,
        causes: Vec<EventId>,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.emit(source, &format!("intend: {goal}"), conv_id, causes, signer)
    }

    /// Decompose breaks a goal into actionable steps. (Choice + Derive)
    pub fn decompose(
        &mut self,
        source: ActorId,
        subtask: &str,
        goal: EventId,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.derive(source, &format!("decompose: {subtask}"), goal, conv_id, signer)
    }

    /// Assign gives work to a specific actor. (Commitment + Delegate)
    pub fn assign(
        &mut self,
        source: ActorId,
        assignee: ActorId,
        scope: &DomainScope,
        weight: Weight,
        cause: EventId,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.delegate(source, assignee, scope, weight, cause, conv_id, signer)
    }

    /// Claim takes on unassigned work. (Intent + Emit)
    pub fn claim(
        &mut self,
        source: ActorId,
        work: &str,
        causes: Vec<EventId>,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.emit(source, &format!("claim: {work}"), conv_id, causes, signer)
    }

    /// Prioritize ranks work by importance. (Value + Annotate)
    pub fn prioritize(
        &mut self,
        source: ActorId,
        target: EventId,
        priority: &str,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.annotate(source, target, "priority", priority, conv_id, signer)
    }

    /// Block flags work that cannot proceed. (Risk + Annotate)
    pub fn block(
        &mut self,
        source: ActorId,
        target: EventId,
        blocker: &str,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.annotate(source, target, "blocked", blocker, conv_id, signer)
    }

    /// Unblock removes an impediment to work. (Consequence + Emit)
    pub fn unblock(
        &mut self,
        source: ActorId,
        resolution: &str,
        causes: Vec<EventId>,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.emit(source, &format!("unblock: {resolution}"), conv_id, causes, signer)
    }

    /// Progress reports incremental advancement. (Commitment + Extend)
    pub fn progress(
        &mut self,
        source: ActorId,
        update: &str,
        previous: EventId,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.extend(source, &format!("progress: {update}"), previous, conv_id, signer)
    }

    /// Complete marks work as done with evidence. (Consequence + Emit)
    pub fn complete(
        &mut self,
        source: ActorId,
        summary: &str,
        causes: Vec<EventId>,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.emit(source, &format!("complete: {summary}"), conv_id, causes, signer)
    }

    /// Handoff transfers work between actors. (Signal + Consent)
    pub fn handoff(
        &mut self,
        from: ActorId,
        to: ActorId,
        description: &str,
        scope: &DomainScope,
        cause: EventId,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.consent(from, to, &format!("handoff: {description}"), scope, cause, conv_id, signer)
    }

    /// Scope defines what an actor may do autonomously. (Capacity + Delegate)
    pub fn scope(
        &mut self,
        source: ActorId,
        target: ActorId,
        scope: &DomainScope,
        weight: Weight,
        cause: EventId,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.delegate(source, target, scope, weight, cause, conv_id, signer)
    }

    /// Review evaluates completed work. (Consequence + Respond)
    pub fn review(
        &mut self,
        source: ActorId,
        assessment: &str,
        target: EventId,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<Event> {
        self.0.respond(source, &format!("review: {assessment}"), target, conv_id, signer)
    }

    // --- Named Functions (6) ---

    /// StandupResult holds the events produced by a Standup.
    /// Standup gathers status from all participants and sets priority:
    /// Progress (batch) + Prioritize.
    pub fn standup(
        &mut self,
        participants: &[ActorId],
        updates: &[&str],
        lead: ActorId,
        priority: &str,
        causes: &[EventId],
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<StandupResult> {
        if participants.len() != updates.len() {
            return Err(crate::errors::EventGraphError::GrammarViolation {
                detail: "standup: participants and updates must have equal length".to_string(),
            });
        }

        let mut result_updates = Vec::new();
        let mut last_id: Option<EventId> = None;
        for (i, actor) in participants.iter().enumerate() {
            let prev = if i == 0 {
                causes.first().cloned().ok_or_else(|| {
                    crate::errors::EventGraphError::GrammarViolation {
                        detail: "standup: causes must not be empty".to_string(),
                    }
                })?
            } else {
                last_id.clone().unwrap()
            };
            let progress = self.progress(actor.clone(), updates[i], prev, conv_id.clone(), signer)?;
            last_id = Some(progress.id.clone());
            result_updates.push(progress);
        }

        let prio = self.prioritize(lead, last_id.unwrap(), priority, conv_id, signer)?;

        Ok(StandupResult {
            updates: result_updates,
            priority: prio,
        })
    }

    /// Retrospective reviews work and identifies improvements: Review (batch) + Intend.
    pub fn retrospective(
        &mut self,
        reviewers: &[ActorId],
        assessments: &[&str],
        lead: ActorId,
        improvement: &str,
        target: EventId,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<RetrospectiveResult> {
        if reviewers.len() != assessments.len() {
            return Err(crate::errors::EventGraphError::GrammarViolation {
                detail: "retrospective: reviewers and assessments must have equal length".to_string(),
            });
        }

        let mut reviews = Vec::new();
        let mut review_ids = Vec::new();
        for (i, reviewer) in reviewers.iter().enumerate() {
            let rev = self.review(reviewer.clone(), assessments[i], target.clone(), conv_id.clone(), signer)?;
            review_ids.push(rev.id.clone());
            reviews.push(rev);
        }

        let improvement_ev = self.intend(lead, improvement, review_ids, conv_id, signer)?;

        Ok(RetrospectiveResult {
            reviews,
            improvement: improvement_ev,
        })
    }

    /// Triage prioritises and assigns a batch of items: Prioritize + Assign + Scope (batch).
    pub fn triage(
        &mut self,
        lead: ActorId,
        items: &[EventId],
        priorities: &[&str],
        assignees: &[ActorId],
        scopes: &[DomainScope],
        weights: &[Weight],
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<TriageResult> {
        let n = items.len();
        if priorities.len() != n || assignees.len() != n || scopes.len() != n || weights.len() != n {
            return Err(crate::errors::EventGraphError::GrammarViolation {
                detail: "triage: all slices must have equal length".to_string(),
            });
        }

        let mut result = TriageResult {
            priorities: Vec::new(),
            assignments: Vec::new(),
            scopes: Vec::new(),
        };

        for i in 0..n {
            let prio = self.prioritize(lead.clone(), items[i].clone(), priorities[i], conv_id.clone(), signer)?;
            let assign = self.assign(lead.clone(), assignees[i].clone(), &scopes[i], weights[i], prio.id.clone(), conv_id.clone(), signer)?;
            let scope_ev = self.scope(lead.clone(), assignees[i].clone(), &scopes[i], weights[i], assign.id.clone(), conv_id.clone(), signer)?;
            result.priorities.push(prio);
            result.assignments.push(assign);
            result.scopes.push(scope_ev);
        }

        Ok(result)
    }

    /// Sprint plans a work cycle: Intend + Decompose + Assign (batch).
    pub fn sprint(
        &mut self,
        source: ActorId,
        goal: &str,
        subtasks: &[&str],
        assignees: &[ActorId],
        scopes: &[DomainScope],
        causes: Vec<EventId>,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<SprintResult> {
        if subtasks.len() != assignees.len() || subtasks.len() != scopes.len() {
            return Err(crate::errors::EventGraphError::GrammarViolation {
                detail: "sprint: subtasks, assignees, and scopes must have equal length".to_string(),
            });
        }

        let intent = self.intend(source.clone(), goal, causes, conv_id.clone(), signer)?;

        let mut result_subtasks = Vec::new();
        let mut assignments = Vec::new();
        for (i, st) in subtasks.iter().enumerate() {
            let task = self.decompose(source.clone(), st, intent.id.clone(), conv_id.clone(), signer)?;
            let assign = self.assign(
                source.clone(),
                assignees[i].clone(),
                &scopes[i],
                Weight::new(0.5).unwrap(),
                task.id.clone(),
                conv_id.clone(),
                signer,
            )?;
            result_subtasks.push(task);
            assignments.push(assign);
        }

        Ok(SprintResult {
            intent,
            subtasks: result_subtasks,
            assignments,
        })
    }

    /// Escalate moves stuck work up: Block + Handoff (to higher authority).
    pub fn escalate(
        &mut self,
        source: ActorId,
        blocker: &str,
        task: EventId,
        authority: ActorId,
        scope: &DomainScope,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<EscalateResult> {
        let block_ev = self.block(source.clone(), task, blocker, conv_id.clone(), signer)?;
        let handoff_ev = self.handoff(source, authority, blocker, scope, block_ev.id.clone(), conv_id, signer)?;

        Ok(EscalateResult {
            block_event: block_ev,
            handoff_event: handoff_ev,
        })
    }

    /// DelegateAndVerify is a full delegation cycle: Assign + Scope.
    pub fn delegate_and_verify(
        &mut self,
        source: ActorId,
        assignee: ActorId,
        scope: &DomainScope,
        weight: Weight,
        cause: EventId,
        conv_id: ConversationId,
        signer: &dyn Signer,
    ) -> Result<DelegateAndVerifyResult> {
        let assign = self.assign(source.clone(), assignee.clone(), scope, weight, cause, conv_id.clone(), signer)?;
        let scope_ev = self.scope(source, assignee, scope, weight, assign.id.clone(), conv_id, signer)?;

        Ok(DelegateAndVerifyResult {
            assign_event: assign,
            scope_event: scope_ev,
        })
    }
}

pub struct StandupResult {
    pub updates: Vec<Event>,
    pub priority: Event,
}

pub struct RetrospectiveResult {
    pub reviews: Vec<Event>,
    pub improvement: Event,
}

pub struct TriageResult {
    pub priorities: Vec<Event>,
    pub assignments: Vec<Event>,
    pub scopes: Vec<Event>,
}

pub struct SprintResult {
    pub intent: Event,
    pub subtasks: Vec<Event>,
    pub assignments: Vec<Event>,
}

pub struct EscalateResult {
    pub block_event: Event,
    pub handoff_event: Event,
}

pub struct DelegateAndVerifyResult {
    pub assign_event: Event,
    pub scope_event: Event,
}
